docs/source/tutorials/arch_modeling/Spypads_tutorial.rst

```diff
@@ -0,0 +1,334 @@
+
+The waveforms will appear similar to :numref:`fig_spypad_waves`
+
+.. _fig_spypad_waves:
+
+.. figure:: ./figures/spypad_waveforms.png
+   :scale: 100%
+
+   Spypad Waveforms
+
+Building Spypads
+~~~~~~~~~~~~~~~~
+
+We will modify the ``k6_frac_N10_adder_chain_40nm_openfpga.xml`` file found in OpenFPGA to expose the **sumout** output from the **ADDF** module. We can start modifying
+the file by running the following command:
+
+.. code-block:: bash
+
+   emacs openfpga_flow/openfpga_arch/k6_frac_N10_adder_chain_40nm_openfpga.xml
+
+Replace **LINE214** with the following:
+
+.. code-block:: xml
+
+   <port type="output" prefix="sumout" lib_name="SUM" size="1" is_global=”true” is_io=”true”/>
+
+**sumout** is now a global output. **sumout** will show up in the ``fpga_top.v`` file and will have waveforms in GTKWave if we run the **full testbench**. To run the 
+**full testbench**, we have to modify the ``hard_adder`` configuration file:
+
+.. code-block:: bash
+
+   emacs openfpga_flow/tasks/fpga_verilog/adder/hard_adder/config/task.conf
+
+Comment out the last line of the file to run the **full testbench**:
+
+.. code-block:: python
+
+   #vpr_fpga_verilog_formal_verification_top_netlist=
+
+We now run the task to see our changes:
+
+.. code-block:: bash
+
+   python3 openfpga_flow/scripts/run_fpga_task.py fpga_verilog/adder/hard_adder --debug --show_thread_logs
+
+We can view the global ports in ``fpga_top.v`` by running the following command:
+
+.. code-block:: bash
+
+   emacs openfpga_flow/tasks/fpga_verilog/adder/hard_adder/run064/k6_frac_N10_tileable_adder_chain_40nm/and2/MIN_ROUTE_CHAN_WIDTH/SRC/fpga_top.v
+
+The ``fpga_top.v`` should have the following in its module definition:
+
+.. code-block:: verilog
+
+   module fpga_top(pReset,
+                prog_clk,
+                set,
+                reset,
+                clk,
+                gfpga_pad_ADDF_sumout,
+                gfpga_pad_GPIO_PAD,
+                ccff_head,
+                ccff_tail);
+   //----- GLOBAL PORTS -----
+   input [0:0] pReset;
+   //----- GLOBAL PORTS -----
+   input [0:0] prog_clk;
+   //----- GLOBAL PORTS -----
+   input [0:0] set;
+   //----- GLOBAL PORTS -----
+   input [0:0] reset;
+   //----- GLOBAL PORTS -----
+   input [0:0] clk;
+   //----- GPOUT PORTS -----
+   output [0:19] gfpga_pad_ADDF_sumout;
+
+We can view the waveform by running GTKWave:
+
+.. code-block:: bash
+
+   gtkwave openfpga_flow/tasks/fpga_verilog/adder/hard_adder/latest/k6_frac_N10_tileable_adder_chain_40nm/and2/MIN_ROUTE_CHAN_WIDTH/and2_formal.vcd &
+
+The waveform should have some changes to its value. An example of what it may look like is displayed in :numref:`fig_spy_adder`
+
+.. _fig_spy_adder:
+
+.. figure:: ./figures/spyadder_waveform.png
+   :scale: 100%
+
+   Sumout Waveform
+
+Conclusion
+~~~~~~~~~~
+
+In this tutorial, we have shown how to build spypads into OpenFPGA Architectures using XML Syntax. If you have any issues, feel free to `contact`_ us.

```

Comment:
If the link is inside the openfpga documentation, please do not use URL. Instead, use reference, e..g,  :ref:`contact`